src/librustc_mir/transform/generator.rs

```diff
@@ -0,0 +1,777 @@
+    };
+
+    // Make sure we remove dead blocks to remove
+    // unrelated code from the resume part of the function
+    simplify::remove_dead_blocks(mir);
+
+    dump_mir(tcx, None, "generator_drop", &0, source, mir);
+}
+
+fn insert_resume_after_return<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                        def_id: DefId,
+                                        mir: &mut Mir<'tcx>) -> Option<BasicBlock> {
+    let drop_arg = mir.local_decls.raw[2].ty.needs_drop(tcx, tcx.param_env(def_id));
+
+    let cleanup = if drop_arg {
+        Some(BasicBlock::new(mir.basic_blocks().len() + 1))
+    } else {
+        None
+    };
+
+    let assert_block = BasicBlock::new(mir.basic_blocks().len());
+    let term = TerminatorKind::Assert {
+        cond: Operand::Constant(box Constant {
+            span: mir.span,
+            ty: tcx.types.bool,
+            literal: Literal::Value {
+                value: ConstVal::Bool(false),
+            },
+        }),
+        expected: true,
+        msg: AssertMessage::GeneratorResumedAfterReturn,
+        target: assert_block,
+        cleanup: cleanup,
+    };
+
+    let source_info = SourceInfo {
+        span: mir.span,
+        scope: ARGUMENT_VISIBILITY_SCOPE,
+    };
+
+    mir.basic_blocks_mut().push(BasicBlockData {
+        statements: Vec::new(),
+        terminator: Some(Terminator {
+            source_info,
+            kind: term,
+        }),
+        is_cleanup: false,
+    });
+
+    if drop_arg {
+        let resume_block = BasicBlock::new(mir.basic_blocks().len() + 1);
+
+        let term = TerminatorKind::Drop {
+            location: Lvalue::Local(Local::new(2)),
+            target: resume_block,
+            unwind: None,
+        };
+
+        mir.basic_blocks_mut().push(BasicBlockData {
+            statements: Vec::new(),
+            terminator: Some(Terminator {
+                source_info,
+                kind: term,
+            }),
+            is_cleanup: true,
+        });
+
+        mir.basic_blocks_mut().push(BasicBlockData {
+            statements: Vec::new(),
+            terminator: Some(Terminator {
+                source_info,
+                kind: TerminatorKind::Resume,
+            }),
+            is_cleanup: true,
+        });
+    }
+
+    cleanup
+}
+
+fn generate_resume<'a, 'tcx>(
+                tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                mut transform: TransformVisitor<'a, 'tcx>,
+                node_id: NodeId,
+                def_id: DefId,
+                source: MirSource,
+                cleanup: Option<BasicBlock>,
+                mir: &mut Mir<'tcx>) {
+    // Poison the generator when it unwinds
+    for block in mir.basic_blocks_mut() {
+        let source_info = block.terminator().source_info;
+        if let &TerminatorKind::Resume = &block.terminator().kind {
+            block.statements.push(transform.set_state(1, source_info));
+        }
+    }
+
+    let source_info = SourceInfo {
```

Comment:
The rest of the function should probably be called `generate_entry_point`, or have a big comment that explains that "resume" means "entry point" rather than `TerminatorKind::Resume`.